custom_profile_fields: Add 'Select multiple' field type

PedroSimas4 opened the pull request:
This PR adds support for a new custom profile field type: SELECT_MULTIPLE (ID 9). This allows administrators to define a list of fixed options (similar to the standard Select field), but enables users to select multiple answers via checkboxes.

Key Changes:

Backend: Added SELECT_MULTIPLE to models.py, added validation logic in validator.py (ensuring values are valid lists of option keys), and updated users.py to handle storage.
Frontend: Updated settings UI to allow creating these fields and the Profile UI to display selected values as a comma-separated list.
Storage: Data is stored in the database as a JSON-encoded list of strings (e.g., ["1", "2"]).

Fixes: #17495

How changes were tested:

Automated Tests:

Added backend tests in zerver/tests/test_users.py to verify creation, validation (invalid choices, malformed JSON), and storage logic.
Added frontend unit tests in web/tests/settings_profile_fields.test.cjs to verify the UI configuration logic.
Ran ./tools/test-all locally and all tests passed.

Manual Verification:

Verified the end-to-end flow:
Created a "Programming Languages" multi-select field in Organization Settings.
Edited a user profile to select multiple options (e.g., Python, Rust).
Verified that the profile displays "Python, Rust" correctly.
Verified that deselecting all options hides the field from the profile display.

Screenshots and screen captures:

<img width="593" height="645" alt="{0D6E12F4-B2CA-437D-9B79-9FD29ECF099A}" src="https://github.com/user-attachments/assets/e20e2fde-37bd-46ae-8db7-0ce59338b00d" />

<img width="591" height="765" alt="{D52474B0-DF8F-4AA0-ADB9-D85EF72EEB93}" src="https://github.com/user-attachments/assets/01bf6848-1e29-41c9-8ccf-833daab84100" />

<img width="873" height="650" alt="{AA6C4E92-4109-457C-B8D8-BD2E3670A905}" src="https://github.com/user-attachments/assets/f8eed74b-56f7-489e-8d62-ba799eb8a3b4" />

<img width="878" height="1244" alt="{036A4A28-9DF5-4B11-A73F-E732BC99C2C8}" src="https://github.com/user-attachments/assets/cee9f28f-4fb9-4898-8bc0-b4c5ef4cae83" />

<img width="588" height="888" alt="{A5EF9FD2-7A60-4FDE-9E23-25EC489261AE}" src="https://github.com/user-attachments/assets/ef346af7-8341-4fff-ac3c-7ffdbb93ab6a" />

Self-review checklist

[x] [Self-reviewed]

Communicate decisions, questions, and potential concerns.

[x] Explains differences from previous plans (e.g., issue description).
[x] Highlights technical choices and bugs encountered.
[x] Calls out remaining decisions and concerns.
[x] Automated tests verify logic where appropriate.

Individual commits are ready for review.

[x] Each commit is a coherent idea.
[x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

[x] Visual appearance of the changes.
[x] Responsiveness and internationalization.
[x] Strings and tooltips.
[x] End-to-end functionality of buttons, interactions and flows.
[x] Corner cases, error conditions, and easily imagined bugs.

PedroSimas4 committed:
custom_profile_fields: Add backend support for multi-select fields.

This commit introduces a new custom profile field type, SELECT_MULTIPLE
(ID 9). This field allows administrators to define a fixed set of
options, similar to the existing SELECT type, but enables users to
choose multiple answers.

Key implementation details:

* Storage: Selected values are stored in the database as a JSON-encoded
  list of strings (e.g., '["1", "2"]').

* Validation: Added logic to `validator.py` to ensure that saved
  values are valid lists and that every selected item exists in the
  field's defined options.

* Tests: Added backend tests in `test_users.py` to verify creation,
  validation, and storage logic.

timabbott commented on web/src/user_profile.ts:
```diff
@@ -482,6 +482,33 @@ export function get_custom_profile_field_data(
+                        readable_values.push(option.text);
+                    }
+                }
+                const final_text = readable_values.join(", ");
```
I think you want to use one of the helpers from `util.ts` that use `Intl.ListFormat` to do this in an internationalized fashion. (Japanese doesn't use `, ` as the separator, for example).

timabbott commented on web/src/custom_profile_fields_ui.ts:
```diff
@@ -45,18 +46,41 @@ export function append_custom_profile_fields(element_id: string, user_id: number
+                        selected_values.add(item);
+                    }
+                }
+            }
```
This is a lot of parsing code; can you add comments explaining what possible failures you're worried about here?

alya commented:
On the UI side:

- A screenshot of the dropdown that includes this type of field is missing.
- The spacing between options is incorrect -- there needs to be more space on the left of the checkbox than the right, if we actually want to put them on the same line. I'd also like to see a screenshot that includes surrounding UI context, and has some options with long names. How is line-wrapping handled?

timabbott commented:
This looks like a great first PR @PedroSimas4! FYI [#api design > custom profile fields @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/custom.20profile.20fields/near/2324428) will need to be resolved before we can finalize the server part of the implementation. 

.

PedroSimas4 committed:
custom_profile_fields: Add frontend UI for multi-select fields.

This updates the web application to support the display and editing
of the new 'Select multiple' custom profile field type.

Changes include:

* Settings UI: Admins can now create and edit multi-select fields in
  the organization settings.

* Profile Display: The user profile now renders these fields as a
  comma-separated list of selected values. If no value is selected,
  the field is hidden to maintain a clean UI.

* Tests: Updated `settings_profile_fields.test.cjs` and mocks to
  support the new field type ID.

Fixes #17495.

PedroSimas4 commented:
@timabbott @alya Thank you for the detailed feedback! I have updated the PR addressing all the points:


1. UI/UX & CSS:

Changed the field label to "Multi-select" as suggested.

Moved the inline styles to web/styles/settings.css and cleaned up the HBS template.

Fixed the flexbox wrapping and spacing for long values.

(Added screenshots of the Admin UI and Profile View).

2. Frontend Logic:

Switched to util.format_array_as_list for proper internationalization.

Replaced generic try/catch blocks with specific z.safeParse checks to handle data integrity safely without swallowing logic errors.

3. Backend & Tests:

Removed # nocoverage tags.

Added a regression test case in zerver/tests/test_users.py covering both invalid JSON input and corrupted field definitions.

Refactored the test flow to avoid the unnecessary try/except block for the field type constant.


Ready for another look!

alya commented:
Hm, I still don't see a screenshot of the dropdown?

timabbott commented:
Tagging @sahil839 for maintainer review once the renaming feedback above has been completed.

sahil839 reviewed (commented):
I have only done review for the first commit currently. Will do the complete review tommorrow.

sahil839 commented on web/src/custom_profile_fields_ui.ts:
```diff
@@ -46,18 +47,44 @@ export function append_custom_profile_fields(element_id: string, user_id: number
+                );
+            } catch {
+                field_choice_dict = {};
+            }
```
I don't think we need this try-catch block. We expect server to validate `field_data` when creating and updating the field so we can expect `field.field_data` to always be a valid json.

sahil839 commented on web/src/settings_profile_fields.ts:
```diff
@@ -444,9 +444,17 @@ function set_up_select_field_edit_form(
+    // We wrap this in a try/catch so that if the field data is malformed
+    // (e.g. invalid JSON), the edit form still opens (showing empty choices)
+    // rather than crashing. This allows the admin to recover by re-entering choices.
+    try {
```
Is there any case that you can reproduce with invalid JSON without manually modifying the DB? If not we can remove this as well.

sahil839 reviewed (commented):
Posted some comments in frontend code as well. I feel the try-catch blocks are not needed since we can expect valid JSON value from server where we already validate the data. Or would be good if you could highlight cases where JSON value can be invalid except for modifying the DB value manually.

alya commented:
Thanks for the work here @PedroSimas4 ! Closing in favor of continued work on #38770.
